lib/parsers/pdf_parser.rb

```diff
@@ -0,0 +1,49 @@
+class FormatParser::PDFParser
+  include FormatParser::IOUtils
+
+  class InvalidStructure < StandardError
+  end
+
+  # First 9 byes of a PDF should be in this format, according to:
+  #
+  #  https://stackoverflow.com/questions/3108201/detect-if-pdf-file-is-correct-header-pdf
+  #
+  # There are however exceptions, which are left out for now.
+  #
+  PDF_MARKER = /%PDF-1\.[0-8]{1}/
+
+  # Page counts have different markers depending on
+  # the PDF type. There is not a single common way of solving
+  # this. The only way of solving this correctly is by adding
+  # different types of PDF's in the specs.
+  #
+  COUNT_MARKER = /\/(N|Count|Page)\s([0-9]+)/
+
+  def call(io)
+    io = FormatParser::IOConstraint.new(io)
+
+    return unless safe_read(io, 9) =~ PDF_MARKER
+
+    FormatParser::Document.new(
+      format: :pdf,
+      page_count: page_count(io)
+    )
+  end
+
+  private
+
+  def page_count(io)
+    io.seek(9)
+
+    while read = safe_read(io, 100)
```

Comment:
- `safe_read` will raise an exception if you do not read the exact amount of bytes requested, so if the filesize is not an exact multiple of 100 (+9 to account for the PDF_MARKER), this will blow up if nothing in the document matches the COUNT_MARKER.
- Further, this does a potentially unbounded number of reads. This might be a problem if someone tries to determine the format of `malicious_20_gb_long.pdf`. We should probably have some way to limit the total amount read.
- If the COUNT_MARKER is spread across read chunks, this will not pick it up and we'll keep reading 100 byte chunks until we hit EOF.

Reply:
> safe_read will raise an exception if you do not read the exact amount of bytes requested, so if the filesize is not an exact multiple of 100 

I've seen this problem occur while fiddling around with the code actually :sweat_smile: . The alternative I do see work is to not use `safe_read` and just read the 100 bytes with `io.read(100)`. I'm guessing there are other smarter measures to take especially considering the following:

> Further, this does a potentially unbounded number of reads. This might be a problem if someone tries to determine the format of malicious_20_gb_long.pdf. We should probably have some way to limit the total amount read.

Yes, we should. What is an acceptable amount of reads? 

> If the COUNT_MARKER is spread across read chunks, this will not pick it up and we'll keep reading 100 byte chunks until we hit EOF.

I'm not sure if I fully understand this, you mean if the COUNT_MARKER is concealed within a serie of bytes? Like:

```
AAAAAAAAAAAAAAAA\Count 5
```

Instead of:

```
\Count 5
```

Reply:
If we read in chunks of 100 bytes and the first bytes of the COUNT_MARKER are in the final bytes of chunk 1 and the last bytes are the first bytes of chunk 2, then it will not be picked up by the regex even though there is a valid COUNT_MARKER in the file.

According to the PDF specs, there can be many count markers in a document as well, for example for each chapter in a document. So you'd run the risk of missing the first one and then misreporting the page count of one of the chapters as the count of the whole thing.

I think it'd be reasonable to read the first 512 or so bytes 🤔